Cleanup byte arrays

boaks opened the pull request:
Cleanup byte arrays with secrets..
Use destroyable SecretKey and SecretIvParameterSpec instead of byte
arrays. Make DTLSSession, SessionTikcet, DTLSConnectionState, and Handshaker also destroyable. Introduce DtlsBlockConnectionState and DtlsAeadConnectionState.
Reduce copy of secrets.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

boaks commented:
FMPOV this would round up the recent security update by reducing the visibility of secrets in heap. It changes the `PskStore` API and therefore I would prefer to have that in the 2.0.0 release. I redesigned the `DTLSConnectionState` and `Record` for reducing the number of copies of these secrets, but I also didn't like the repeated "switch (cipherSuite.getCipherType())". 
There is still potential for improvement: moving the sequence numbers out of DTLSSession into DTLSConnectionState, add a factory to the new `SecretUtil`, or adapt the code to use "(H)SM". But that would be too future stories for 2.0.0.  

On the other side, I would understand not to merge 2000 changed lines that short before 2.0.0, and looking closer to the "secrets in heap" shows also, that this is only a slightly improvement.

So, please, give me your feedback, if we should work on this before the 2.0.0, afterwards (but it changes the API), or not at all.

sbernard31 commented:
> this would round up the recent security update by reducing the visibility of secrets in heap

Did I miss something ?

boaks commented:
> Did I miss something ?

Hopefully not. That was the cleanup work in the handshake and AES-CBC cipher. FMPOV, the unspecific usage of byte array may also be considered as future security thread. Therefore I started to see, if I can adapt the code to use `SecretKey` with the `Destroyable` extension of more modern jdk's.

sbernard31 commented:
> FMPOV, the unspecific usage of byte array may also be considered as future security thread. Therefore I started to see, if I can adapt the code to use SecretKey with the Destroyable extension of more modern jdk's.

Do you have any resources to share about this ? 

> On the other side, I would understand not to merge 2000 changed lines that short before 2.0.0, and looking closer to the "secrets in heap" shows also, that this is only a slightly improvement.

My first feeling, that we should maybe plan this after the 2.0.0.

boaks commented:
> Do you have any resources to share about this ?

[Java 8 , HmacCore](https://github.com/frohoff/jdk8u-dev-jdk/blob/master/src/share/classes/com/sun/crypto/provider/HmacCore.java#L95)

As you can see, newer (than 7) java implementations care about cleanup byte arrays with secret. Newer java also added the "Destroyable" interface.
The reason behind that seems to be, that "dump" may otherwise contain secrets in "plain". The usage of `SecretKey`itself only reduces the amount of copies of such byte arrays filled with secrets. But with the changes n this PR it will get much easier to replace it by other implementations (maybe in trade of performance).

Many lines are just replacing "byte[]" by "SecretKey". The second commit introduces the specific Dtls??ConenctionState, which I think, is also worth. The risk by these changes are not to high, though tests and benchmarks validate, that it works.

For using applications the change would be to start using also a "SecretKey" instead of "plain byte arrays". Even if I don't insist in this PR, I would prefer to have it in 2.0.0.

boaks commented on element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramWriter.java:
```diff
@@ -40,8 +39,24 @@
+		byteStream = secureClose ? new ByteArrayOutputStream() {
+
+			@Override
+			public void close() throws IOException {
```
OK.

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/util/SecretUtil.java:
```diff
@@ -0,0 +1,232 @@
+		if (clean != null) {
+			try {
+				clean.destroy();
+			} catch (DestroyFailedException e) {
```
DEBUG?
I think, it should be tested during development, if it fits the expectations and requirements.

sbernard31 replied:
With events which should never happened, I think raise a WARN which explain that should never happen is pretty much a good practice.

boaks replied:
FMPOV there are two case:

- This should never happen, but it would otherwise harm seriously
- This should never happen, and it's simply not worth to spend too much time in it.

For me it's simply the second. But though switch to warn would save also time, why not.

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/util/SecretUtil.java:
```diff
@@ -0,0 +1,232 @@
+	 */
+	public static SecretKey create(SecretKey key) {
+		SecretKey result = null;
+		if (key != null) {
```
The idea, pattern, design:

Always destroy the "secrets" you maintain.
Therefore always copy passed in arguments, if you keep them after returning.
Always copy returned secret.
If you get a secret, destroy it after your ready.

Other ideas would be to document some "special house keeping" plans, but with the above it gets simpler in my opinion.

sbernard31 replied:
This sounds to generate a lot of copy, but I agree this is probably less error prone.

I would let more specific comment in my next review.

boaks replied:
FMPOV, this generates more copies during handshake, but cleans them up afterwards in time. After the handshake, the redesign of the DtlsConnectionState and moving encrypt/decrypt from Records switch into the special subtypes of the DtlsConnectionState using the SecretKey direct in the own scope of that specific DtlsConnectionState without get, doesn't create more copies. But ensures, that the key are destroyed with the session.

boaks commented on element-connector/src/main/java/org/eclipse/californium/elements/util/DatagramWriter.java:
```diff
@@ -201,10 +216,26 @@ public void write(DatagramWriter data) {
+	public void close() {
+		try {
+			byteStream.close();
+		} catch (IOException e) {
```
In my opinion nothing happens, if such an exception is thrown.
But if you want, I can add a INFO (or WARN).

sbernard31 replied:
With events which should never happened, I think raise a WARN which explain that should never happen is pretty much a good practice.

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/StringArrayPskStore.java:
```diff
@@ -0,0 +1,133 @@
+ * <p>
+ * Only to be used for backwards compatibility!
+ */
+public abstract class StringArrayPskStore implements PskStore {
```
I want to give an hint, how to migrate.
I think deprecate would be good, and remove for the 2.0.0.

boaks commented:
1. yes. I thought already on this. But as you see, it's 2500 lines.
And that behavior is not used and therefore could be added later just as implementation  update. But I can add also some more line, if you prefer it now.

2. the create is called to either keep the secret after returning a function or when secrets are returned. In both cases the new one should be "Destroyable", if the intention is to implement "house-keeping". Yes, if the origin is "insecure", this may be somehow in vain, but at least, it's not too bad.

3. I already confessed

> and looking closer to the "secrets in heap" shows also, that this is only a slightly improvement.

The new solution, doesn't limit the number of copies, it's more the time they contain the secret. Only clear the returned byte arrays would leak the secrets maintained by older "SecureKey" implementation after the time they used. This PR cleans up all byte arrays and all "secret object" (maybe except those you found left over :-) ).

boaks commented:
3. 
> DtlsBlockConnectionState and DtlsAeadConnectionState.
> Reduce copy of secrets.

That reduction is something the first commit introduced by returning copies.
With the new ConnectionState, the copies are reduced.

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java:
```diff
@@ -1715,15 +1717,16 @@ private void resumeExistingSession(ClientHello clientHello, Record record, final
 				}
 			}
+			if (!ok && ticket != null) {
+				SecretUtil.destroy(ticket);
```
Because of the exception.

sbernard31 replied:
OK

boaks commented:
I addressed some of your comments.
So the main open question for me would be, if I should change the "Destroyable" implementations throwing exception on usage after being destroyed.
I would go for implementation, where the "Secret??" throws that exception, and containing objects when accessing that secrets.

boaks commented:
I added the `IllegalStateExceptions` to the secret objects.
I would like to squash this PR, in order to have not code changed twice or more.

boaks commented:
I would like to include this in 2.0.0-M18 / RC1

sbernard31 commented:
I started to review this again but most of my comments are about  duplicating bytes...  which makes me think that this will not be really useful...

This PR introduce an API improvement.
My feeling is there is a lot of avoidable "bytes duplicating" but this is intentional.
Anyway, the benefits of removing this duplication are really hypothetical and keep them is probably less error prone, so let's go with this.

Not directly linked to this PR, but what about PrivateKey and the way we keep them in Memory ?

sbernard31 commented:
> It's not really about copies, it's more about copies left uncontrolled after usage

I understand the meaning of this PR, but I **was** disturb by byte duplications which are done in order to respect : https://github.com/eclipse/californium/pull/1087#discussion_r331541311

> Exactly that controlled clear/destroy reduces the time, when such copies are in heap.

I think some duplicate could be avoided keeping in mind this goal (reduces the time).
(e.g. do not Arrays.copy when you create a DestroyableSecretKeySpec from full byte array)
But this will break rules you defined(https://github.com/eclipse/californium/pull/1087#discussion_r331541311). And breaking this rules is maybe more error prone. 
(and you reduce a lot of this copy with some of your last changes)

That's why I said  "ok let's go like this". :)

boaks commented:
In the meantime I redesigned/cleaned up the PSK processing also. 
That reduces the copies also a few and removes old code duplications :-).

sbernard31 commented:
:+1:

boaks commented:
@sbernard31 

Is it OK to merge it? Or have I overseen an comment?

sbernard31 commented:
LGTM
